File: AdventOfCode2020/Day5.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;
int BinarySearch(string &input, int lowRange, int highRange);
int BinarySearch2(string& input, int lowRange, int highRange);
int main()
{
	bool arr[128][8] = { false };
	int row, column, high = 0;
	string input;
	ifstream infile;
	int counter = 0;
	infile.open("E:\\AdventOfCode\\AdventOfCode2020\\Day5.txt");
	while (!infile.eof())
	{
		getline(infile, input);
		counter++;
		row = BinarySearch(input, 0, 127);
		column = BinarySearch2(input, 0, 7);
		arr[row][column] = true;
		if (high < (row * 8 + column))
			high = row * 8 + column;
	}
	for (int x = 1; x < 7; x++)
	{
		for (int j = 0; j < 128; j++)
		{
			if (!arr[j][x] && arr[j][x-1]&&arr[j][x + 1])
			{
				cout << j * 8 + x << endl;
			}
		}
	}


}
int BinarySearch(string &input, int lowRange, int highRange) {
	if (input.empty())
		return 0;
	else if (input[0] == 'L' || input[0] == 'R')
		return lowRange;
	else if((input[0] == 'F') || (input[0] == 'B'))
	{
		if (input[0] == 'F')// lower half
		{
			input.erase(input.begin());
			
			// 0-127--> 0,63
			// 44-45-->44
			// 
			return BinarySearch(input, lowRange, (highRange + lowRange)/2);
		}
		else if (input[0] == 'B')// higher half
		{
			// 0-63--> 32,63
			// 44-45-->45,45
			input.erase(input.begin());
			return BinarySearch(input, ((highRange+1 +lowRange) / 2), highRange);
		}
	}
}
int BinarySearch2(string& input, int lowRange, int highRange) {
	if (input.empty())
		return lowRange;
	if (input[0] == 'L')// lower half
	{
		input.erase(input.begin());

			// 0-127--> 0,63
			// 44-45-->44
			// 
		return BinarySearch2(input, lowRange, (highRange + lowRange) / 2);
	}
	else if (input[0] == 'R')// higher half
	{
			// 0-63--> 32,63
			// 44-45-->45,45
			input.erase(input.begin());
			return BinarySearch2(input, ((highRange + 1 + lowRange) / 2), highRange);// could cause error
	}
	
}

File: AdventOfCode2020/Day5.py
def main():
    with open("Day5.txt") as f:
        high=0
        a = [[False for x in range(8)] for x in range(128)]
        for x in f:
            x=x.strip()
            row=BinarySearch(x[:7], 0, 127)
            column=BinarySearch2(x[7:],0,7)
            a[row][column]=True
            if (high<(row*8+column)):
                high=row*8+column
        for x in range(len(a)-1):
            for y in range(1,(len(a[x])-1),1): #y is column
                if a[x][y]==False and a[x][y+1]==True and a[x][y-1]==True:
                    print(x*8+y)
        



def BinarySearch(x, low:int, high:int):
    if(len(x)==0):
        return low
    elif(x[0]=="F"):#lower half
        return BinarySearch(x[1:],low, ((low+high)//2))
    elif(x[0]=="B"):
        return BinarySearch(x[1:],((high+1 +low) // 2),high )
def BinarySearch2(x, low:int , high:int):
    if(len(x)==0):
        return low
    elif(x[0]=="L"):#lower half
        return BinarySearch2(x[1:],low, ((low+high)//2))
    elif(x[0]=="R"):
        return BinarySearch2(x[1:],((high+1 +low) // 2),high )


if __name__ == "__main__":
    main()